src/io/parquet/write/mod.rs

```diff
@@ -163,8 +166,31 @@ pub fn array_to_pages(
                     )
                 })
             }
-            _ => array_to_page(array, type_, nested, options, encoding)
-                .map(|page| DynIter::new(std::iter::once(Ok(page)))),
+            _ => {
+                const DEFAULT_PAGE_SIZE: usize = 1024 * 1024;
+                let page_size = options.data_pagesize_limit.unwrap_or(DEFAULT_PAGE_SIZE);
+
+                let bytes_per_row = estimated_bytes_size(array.slice(0, 1).as_ref()).max(1);
+                let row_per_page = page_size / bytes_per_row + 1;
+
+                if array.len() <= row_per_page {
+                    array_to_page(array, type_, nested, options, encoding)
+                        .map(|page| DynIter::new(std::iter::once(Ok(page))))
+                } else {
+                    let first = array.slice(0, row_per_page);
+                    let other = array.slice(row_per_page, array.len() - row_per_page);
+                    Ok(DynIter::new(
+                        array_to_pages(first.as_ref(), type_.clone(), nested, options, encoding)?
```

Comment:
here may reach stackoverflow of recursion.

But if I change to 
```
 _ => {
                const DEFAULT_PAGE_SIZE: usize = 1024 * 1024;
                let page_size = options.data_pagesize_limit.unwrap_or(DEFAULT_PAGE_SIZE);
                let bytes_per_row = ((array_byte_size as f64) / (array.len() as f64)) as usize;
                let row_per_page = (page_size / bytes_per_row).max(1);
                
                let iter = (0..array.len()).step_by(row_per_page).map(move |offset| {
                    let length = if offset + row_per_page > array.len() {
                        array.len() - offset
                    } else {
                        row_per_page
                    };
                    
                    let sub_array = array.slice(offset, length);
                    array_to_page(sub_array.as_ref(), type_, nested, options, encoding)
                });
                
                Ok(DynIter::new(iter))
            }
```

may have lifetime error.

```
 Ok(DynIter::new(iter))
    |                 ^^^^^^^^^^^^^^^^^^^^^^ returning this value requires that `'3` must outlive `'static`
```

@jorgecarleitao  any advice about this?

Reply:
I choose to collect the iter.